File: ch02/2.3/Quick3.java
import java.util.Random;

public class Quick3 {
    private static void exch(Comparable[] a, int i, int j) {
        Comparable swp = a[i];
        a[i] = a[j];
        a[j] = swp;
    }
    private static void sort(Comparable[] a, int lo, int hi) {
        if (hi <= lo) return;

        Comparable v = a[lo];
        int i = lo, lt = lo, gt = hi;
        while (i <= gt) {
            int cmp = a[i].compareTo(v);
            if      (cmp < 0) exch(a, lt++, i++);
            else if (cmp > 0) exch(a, i, gt--);
            else         i++;
        }
        sort(a, lo, lt-1);
        sort(a, gt+1, hi);
    }
    public static void sort(Comparable[] a) {
        StdRandom.shuffle(a);
        sort(a, 0, a.length-1);
    }
    public static void main(String[] args) {
        int N = Integer.parseInt(args[0]);
        Random rgen = new Random();

        Integer[] a = new Integer[N];
        for (int i = 0; i < N; i++) {
            a[i] = rgen.nextInt(5);
        }
        Quick3.sort(a);
        for (int i = 0; i < N; i++) {
            System.out.format("%d ", a[i]);
        }
        System.out.println("");
    }
}

File: ch02/2.3/Quick3.cpp
#include <algorithm>
#include <iostream>
#include <random>
#include <vector>

using namespace std;

void sort(vector<int> &a, size_t lo, size_t hi) {
    // Return condition
    if (hi <= lo)
        return;

    size_t lt = lo, gt = hi, i = lo;
    int v = a[lo];
    while (i <= gt) {
        if (v > a[i])      swap(a[i++], a[lt++]);
        else if (v < a[i]) swap(a[i], a[gt--]);
        else               ++lt;
    }

    sort(a, lo, lt-1);
    sort(a, gt+1, hi);
}

void sort(vector<int> &a) {
    // Shuffle the vector
    auto engine = default_random_engine {};
    shuffle(begin(a), end(a), engine);

    // sort the array
    sort(a, 0, a.size()-1);
}

int main() {
    return 0;
}
